DEVX-2467: Create consumer lag client use case (#948)
confluentinc/examples: ccloud-monitoring/docs/consumer-scenarios/consumer-lag-problem.rst

.. _ccloud-monitoring-consumer-connectivity-problem:

Consumer Lag Problem
********************

> Consumer group lag is a tremendous performance indicator. It tells you the difference between the

Reviewer: ```suggestion
Consumer lag is a tremendous performance indicator. It tells you the offset difference between the
```

producer's last produced message and the consumer group's last commit. If you are unfamiliar with
consumer groups or concepts like committing offsets, please refer to this
`Kafka Consumer documentation <https://docs.confluent.io/platform/current/clients/consumer.html>`__.

> A large or quickly growing lag indicates that the consumer is not able to keep up with

Reviewer: ```suggestion
A large consumer lag, or a quickly growing lag, indicates that the consumer is not able to keep up with
```

the volume of messages on a topic.

> This scenario will look at Confluent Cloud metrics from the Metrics API, `kafka-lag-exporter <https://github.com/lightbend/kafka-lag-exporter>`__ metrics, and

Reviewer: Please use `|ccloud|` instead of `Confluent Cloud`, here and wherever else is applicable

Author: Idk why that is such a hard one for me to remember 😓  will correct throughout these additions

Author: corrected in latest commit f3f5b8f

client metrics from the client application’s MBean object ``kafka.consumer:type=consumer-fetch-manager-metrics,client-id=<client_id>``.

Introduce failure scenario
^^^^^^^^^^^^^^^^^^^^^^^^^^

> #. Stop ``consumer-1`` container, thus removing a consumer from the consumer group and add 2 producers:

Reviewer: - Presumably this command stops ``consumer-1`` because the scale is dropped from 2 to 1, but maybe it's worth an explanation.  I had to do a double-take on this to understand that.
- Add `2` producers?
- Maybe it's worth a reminder to say, something like, by default, X producer and Y consumers are running.  Adjust that to X' and Y' in order to show...

Author: 👍

   Which will produce the following output:

   .. code-block:: bash

      ccloud-exporter is up-to-date
      kafka-lag-exporter is up-to-date
      node-exporter is up-to-date
      grafana is up-to-date
      prometheus is up-to-date
      Stopping and removing ccloud-monitoring_consumer_2 ... done
      Starting ccloud-monitoring_producer_1              ... done
      Creating ccloud-monitoring_producer_2              ... done
      Creating ccloud-monitoring_producer_3              ... done
      Creating ccloud-monitoring_producer_4              ... done
      Creating ccloud-monitoring_producer_5              ... done
      Starting ccloud-monitoring_consumer_1              ... done

Diagnose the problem
^^^^^^^^^^^^^^^^^^^^

#. Open `Grafana <localhost:3000>`__ and login with the username ``admin`` and password ``password``.

#. Navigate to the ``Consumer Client Metrics`` dashboard. Wait 2 minutes and then observe:

   - ``Rebalance rate`` (``rebalance-rate-per-hour``) has a bump, indicating that the consumer group ``demo-cloud-monitoring-1`` underwent a rebalance, which is to be expected when a consumer leaves the group.

   |Consumer Rebalance Bump|

>    - An upward trend in ``Consumer group lag in records``.  ``Consumer group lag in seconds`` will have a less dramatic increase. Both indicating that the producer is creating more messages than the consumer can fetch in a timely manner. These metrics are derived from ``kafka-lag-exporter``.

Reviewer: If ``kafka-lag-exporter`` is not an MBean, explain what is providing those stats (e.g. explain the docker container, what it does, how it works, etc -- not verbose, just a sentence or two)

Author: 👍

      demo-cloud-monitoring-1 demo-topic-1    3          91212           91278           66              consumer-demo-cloud-monitoring-1-1-b0bec0b5-ec84-4233-9d3e-09d132b9a3c7 /10.2.10.251    consumer-demo-cloud-monitoring-1-1
      demo-cloud-monitoring-1 demo-topic-1    4          47854           47893           39              consumer-demo-cloud-monitoring-1-1-b0bec0b5-ec84-4233-9d3e-09d132b9a3c7 /10.2.10.251    consumer-demo-cloud-monitoring-1-1
      demo-cloud-monitoring-1 demo-topic-1    5          47748           47803           55              consumer-demo-cloud-monitoring-1-1-b0bec0b5-ec84-4233-9d3e-09d132b9a3c7 /10.2.10.251    consumer-demo-cloud-monitoring-1-1
      demo-cloud-monitoring-1 demo-topic-1    1          48097           48151           54              consumer-demo-cloud-monitoring-1-1-b0bec0b5-ec84-4233-9d3e-09d132b9a3c7 /10.2.10.251    consumer-demo-cloud-monitoring-1-1
      demo-cloud-monitoring-1 demo-topic-1    2          48310           48370           60              consumer-demo-cloud-monitoring-1-1-b0bec0b5-ec84-4233-9d3e-09d132b9a3c7 /10.2.10.251    consumer-demo-cloud-monitoring-1-1

>    The downside of this view is the lack of historical context that the ``Consumer Client Metrics`` dashboard provides.

Reviewer: ```suggestion
   This command provides a snapshot in time, but it lacks of historical context that the ``Consumer Client Metrics`` dashboard provides.
```

> #. Another view of consumer lag can be found in |ccloud|. Open the UI, navigate to the "Consumers" section and click on the ``demo-cloud-monitoring-1`` consumer group.

Reviewer: Maybe CCloud UI should precede the command line? (swap this with above)

Author: 👍

   This page will update periodically, within two minutes you should see a steady increase is the offset lag.

   |Confluent Cloud Consumer Lag|

   Again the downside of this view is the lack of historical context that the ``Consumer Client Metrics`` dashboard provides.

#. A top level view of the |ccloud| cluster that reflects an increase in bytes produced and bytes consumed can be viewed in the ``Confluent Cloud`` dashboard in the panels highlighted below.

   |Confluent Cloud Request Increase|

> #. The consumer logs won't show that the consumer is falling behind which is why it is important to keep an eye on consumer group metrics.

Reviewer: ```suggestion
#. The consumer logs won't show that the consumer is falling behind which is why it is important to have a robust monitoring solution that covers consumer lag.
```

>    Logs will tell us that the consumer group rebalanced, the logs can be accessed with the following command:

Reviewer: ```suggestion
   However, consumer logs will show that the consumer group rebalanced.  The logs can be accessed with the following command:
```
